### PR TITLE
Concurrent editing notifications front-end

### PR DESCRIPTION
**Pull request by laymonage**
Built on top of #12109 and #12092.

https://github.com/wagtail/wagtail/assets/6379424/de0bab15-3cda-43e1-99e1-3fbbae52528b

### CODE REVIEW

**Comment by laymonage** [comment]
Have now rebased on top of latest main (with #12102 merged) and updated the frontend to use the real endpoints instead of the mock ones. Basic functionalities seem to work, but there are still lots to iron out.

File: wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                        {% avatar user=session.user size="small" %}
+                        <span class="w-text-14">{{ session.user|user_display_name|default:_("system")|capfirst }}</span>
+                    </div>
+                {% endfor %}
+            </div>
+        {% enddropdown %}
+    {% endif %}
+</div>
+
+{% if sessions.0.revision_id %}
```
Comment (laymonage) [inline wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html:98]: This is assuming the user who has the newest `revision_id` is always the first user in the list, but it's OK because the backend does the ordering logic.

File: wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    {% for session in sessions|slice:":4" %}
+        {% if session.revision_id %}
+            {% fragment as avatar %}
+                <div class="w-relative">
+                    {% avatar user=session.user classname="w-mx-[1px] w-w-7 w-h-7 w-border-2 w-border-critical-200 w-ping w-ping--critical" %}
+                    {% icon name="warning" classname="w-text-critical-200 w-w-4 w-h-4 w-absolute -w-bottom-2 w-left-2 w-z-10 w-stroke-surface-header w-shadow-white group-hover:w-hidden" %}
+                </div>
+            {% endfragment %}
+
+            {% blocktranslate trimmed with user_name=session.user|user_display_name|default:_("system") asvar saved_new_version_message %}
```
Comment (laymonage) [inline wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html]: There are cases where the "session" does not really come from an `EditingSession` object. For example, a user saved a new revision since you loaded the editor, but then they left and the session gets cleaned up. In this case, we use information from the revision object to display who the user is.

However, revisions can also be made programmatically without a user attached, in which case we handle it by displaying the user name as "system". We also have a similar handling for log actions/history for when the user is deleted (or if the action was made by system).

File: wagtail/admin/views/generic/mixins.py [hunk 3]
```diff
@@ -710,13 +710,7 @@ def get_editing_sessions(self):
                 "wagtailadmin_editing_sessions:release",
                 args=(session.id,),
             ),
-            list(
```
Comment (laymonage) [inline wagtail/admin/views/generic/mixins.py]: See commit messages for details, but the reason why we don't load the editing sessions immediately on initial render is:
- To prevent race condition with the release endpoint being called when you reload the page i.e. the session is not removed yet when the reload request is sent, causing the previous session to appear initially
- To avoid having to duplicate/extract the logic in the  ping view (annotating the latest revision and sorting the list etc.)

It also helps with performance as it means there's one less query to be performed on initial page render from the server.

File: client/src/controllers/UnsavedController.ts [hunk 4]
```diff
@@ -181,6 +181,9 @@ export class UnsavedController extends Controller<HTMLFormElement> {
     if (!confirmationMessage) return null;
 
     if (this.forceValue || this.hasCommentsValue || this.hasEditsValue) {
+      const confirmEvent = this.dispatch('confirm', { cancelable: true });
+      if (confirmEvent.defaultPrevented) return null;
+
       // eslint-disable-next-line no-param-reassign
       event.returnValue = confirmationMessage;
```
Comment (laymonage) [inline client/src/controllers/UnsavedController.ts:188]: The way we trigger the browser confirmation dialog is actually deprecated. We should've done `event.preventDefault()` instead of setting `returnValue` on the event. See https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event

I wanted so badly to refactor this but I think it's out of scope for this PR... unless you want to review it too

File: client/src/controllers/ActionController.test.js [hunk 5]
```diff
@@ -22,7 +24,17 @@ describe('ActionController', () => {
...
         assign: { configurable: true, value: jest.fn() },
-        reload: { configurable: true, value: jest.fn() },
+        reload: {
+          configurable: true,
+          value: jest.fn().mockImplementation(() => {
+            const event = new Event('beforeunload');
+            Object.defineProperty(event, 'returnValue', {
+              value: null,
+              writable: true,
+            });
```
Comment (laymonage) [inline client/src/controllers/ActionController.test.js:34]: If we also handled the deprecated `returnValue` usage of the `beforeunload` event, then this mock can be simplified by checking whether `event.defaultPrevented` is `true` instead

File: wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html [hunk 6]
```diff
@@ -1,60 +1,62 @@
 {% load wagtailadmin_tags i18n %}
 
-<div class="w-editing-sessions">
+<ol class="w-editing-sessions">
```
Comment (laymonage) [inline wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html:3]: Using `ol` as the list is technically ordered, even if it is not really transparent to the end user

**Review by laymonage — COMMENTED** [review]
Initial comments for the current version of the PR, will add more commits to address a few things we discussed

File: wagtail/admin/templates/wagtailadmin/shared/editing_sessions/module.html [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    data-w-swap-defer-value="true"
+    data-w-action-continue-value="true"
+    data-w-action-url-value="{{ release_url }}"
+    data-w-session-w-dialog-outlet="[data-edit-form] [data-controller='w-dialog']#w-overwrite-changes-dialog"
+    data-action="w-session:ping->w-swap#submit visibilitychange@document->w-session#dispatchVisibilityState w-session:visible->w-session#ping w-session:visible->w-session#addInterval w-session:hidden->w-session#clearInterval w-session:hidden->w-action#sendBeacon w-unsaved:add@document->w-session#setUnsavedChanges w-unsaved:clear@document->w-session#setUnsavedChanges w-swap:json->w-session#updateSessionData"
+>
+    <input type="checkbox" name="is_editing" value="1" data-w-session-target="unsavedChanges" hidden />
+    {% if revision_id %}
+        {# The ID of the latest revision that was known when this module was initially loaded #}
+        <input type="number" name="revision_id" value="{{ revision_id }}" hidden />
```
Comment (thibaudcolas) [inline wagtail/admin/templates/wagtailadmin/shared/editing_sessions/module.html:40]: I’m not clear why this isn’t a `type="hidden"` field 🤔

File: wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        {% if request.user == sessions.0.user %}
+            {% blocktranslate trimmed with model_name=content_type.name asvar someone_has_saved_message %}
+                You have saved a newer version of this {{ model_name }} in another session
+            {% endblocktranslate %}
+        {% else %}
+            {% blocktranslate trimmed with user_name=sessions.0.user|user_display_name|default:_("system") model_name=content_type.name asvar someone_has_saved_message %}
+                {{ user_name }} has saved a newer version of this {{ model_name }}
+            {% endblocktranslate %}
+        {% endif %}
+        {{ someone_has_saved_message|capfirst }}
```
Comment (thibaudcolas) [inline wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html:109]: Wouldn’t `capfirst` interfere with how users wrote their username?

File: wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+        {% if request.user == sessions.0.user %}
+            {% blocktranslate trimmed asvar overwrite_message %}
+                Proceeding will overwrite the changes you made in that session. Refreshing the page will lose any of your unsaved changes in the current session.
+            {% endblocktranslate %}
+        {% else %}
+            {% blocktranslate trimmed with user_name=sessions.0.user|user_display_name|default:_("system") asvar overwrite_message %}
+                Proceeding will overwrite the changes made by {{ user_name }}. Refreshing the page will lose any of your unsaved changes.
+            {% endblocktranslate %}
+        {% endif %}
+        {{ overwrite_message|capfirst }}
```
Comment (thibaudcolas) [inline wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html:122]: I don’t see why `|capfirst` is needed

File: wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+                {{ user_name }} has saved a newer version of this {{ model_name }}
+            {% endblocktranslate %}
+        {% endif %}
+        {{ someone_has_saved_message|capfirst }}
+    </template>
+
+    <template data-controller="w-teleport" data-w-teleport-target-value="#subtitle-w-overwrite-changes-dialog" data-w-teleport-reset-value="true">
+        {% if request.user == sessions.0.user %}
+            {% blocktranslate trimmed asvar overwrite_message %}
+                Proceeding will overwrite the changes you made in that session. Refreshing the page will lose any of your unsaved changes in the current session.
```
Comment (thibaudcolas) [inline wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html:115]: Re the wording – "that session"? Maybe replace with "another session"?

File: wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+            One other user is currently viewing
+        {% plural %}
+            {{ num_user }} other users are currently viewing
+        {% endblocktranslate %}
+
+        {% fragment as more_sessions_toggle %}
+            +{{ sessions|length|add:"-4" }}
+        {% endfragment %}
+
+        {% dropdown theme="drilldown" classname="w-editing-sessions__session w-editing-sessions__session--more" toggle_classname="w-editing-sessions__avatar" toggle_label=more_sessions_toggle toggle_aria_label=num_other_users_message|capfirst %}
```
Comment (thibaudcolas) [inline wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html:85]: I’ve not tested this yet but want to flag we probably shouldn’t use both `toggle_label` and `toggle_aria_label`. It’d be a problem for voice recognition users trying to name the button.

File: client/src/controllers/TooltipController.ts [hunk 12]
```diff
@@ -72,8 +76,14 @@ export class TooltipController extends Controller<HTMLElement> {
   }
 
   get options(): Partial<Props> {
+    let content: Content = this.contentValue;
+    if (this.hasContentTarget) {
+      this.contentTarget.hidden = false;
+      content = this.contentTarget;
+    }
```
[the remark this answers is not in the document]
Comment (thibaudcolas) [reply client/src/controllers/TooltipController.ts:83]: I think this would be worth a brief code comment.

File: client/src/controllers/SwapController.ts [hunk 13]
```diff
@@ -267,10 +270,31 @@ export class SwapController extends Controller<
...
     })
-      .then((response) => {
+      .then(async (response) => {
         if (!response.ok) {
           throw new Error(`HTTP error! Status: ${response.status}`);
         }
+        if (this.jsonPathValue) {
+          let html: unknown;
+          try {
+            const json: Record<string, unknown> = await response.json();
```
[the remark this answers is not in the document]
Comment (thibaudcolas) [reply client/src/controllers/SwapController.ts:295]: I think this could be worth a brief code comment.

File: client/src/controllers/SwapController.ts [hunk 14]
```diff
@@ -262,38 +275,113 @@ export class SwapController extends Controller<
     }) as CustomEvent<{ requestUrl: string }>;
 
     if (beginEvent.defaultPrevented) return Promise.resolve();
-
+    const formMethod = this.formElement.getAttribute('method') || undefined;
```
Comment (thibaudcolas) [inline client/src/controllers/SwapController.ts:278]: Unless it absolutely must be `undefined`?

```suggestion
    const formMethod = this.formElement.getAttribute('method');
```

File: client/src/controllers/SwapController.ts [hunk 15]
```diff
@@ -262,38 +275,113 @@ export class SwapController extends Controller<
...
-        this.dispatch('success', {
-          cancelable: false,
-          detail: { requestUrl, results },
-          target,
-        });
+          return results;
+        };
+
+        // If the currently focused element is within the target container,
+        // defer the write until the focus has left the container.
```
Comment (thibaudcolas) [inline client/src/controllers/SwapController.ts:346]: Upon further testing, I noticed this affects tooltips more than we discussed:

- Hovering over the open tooltip (for example to click a button within) results in it disappearing 
- "Clicking" the tooltip and then moving focus away also results in it disappearing

Is it possible to better manage those scenarios?

File: client/src/controllers/SwapController.ts [hunk 16]
```diff
@@ -197,24 +203,29 @@ export class SwapController extends Controller<
...
-
-    // serialise the form to a query string
-    // https://github.com/microsoft/TypeScript/issues/43797
-    const searchParams = new URLSearchParams(new FormData(form) as any);
+    const form = this.formElement;
+    const data = new FormData(form);
 
+    // serialise the form to a query string if it's a GET request
+    // cast as any to avoid https://github.com/microsoft/TypeScript/issues/43797
+    const searchParams = new URLSearchParams(data as any);
```
Comment (thibaudcolas) [inline client/src/controllers/SwapController.ts:223]: I think it would be clearer if this was done conditionally? As-is the comment is incorrect, the serialisation always happens, it just ends up being unused for other methods

File: client/src/controllers/DropdownController.ts [hunk 17]
```diff
@@ -210,6 +210,12 @@ export class DropdownController extends Controller<HTMLElement> {
           placement: 'bottom',
           plugins: this.plugins,
         },
+        'popup': {
+          arrow: true,
+          placement: 'bottom',
+          plugins: this.plugins,
+          trigger: 'mouseenter focus click',
+        },
```
[the remark this answers is not in the document]
Comment (thibaudcolas) [reply client/src/controllers/DropdownController.ts:218]: I think this is worth a code comment

File: wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html (hunk 10 above)
Comment (laymonage) [reply wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html:115]: This is the subtitle. The title already says "You have saved a newer version of this {{ model_name }} in another session", I think it doesn't read very well if we repeat "another session":

You have saved a newer version of this page in another session
Proceeding will overwrite the changes you made in that session. Refreshing the page will lose any of your unsaved changes in the current session.

File: client/src/controllers/SwapController.ts (hunk 14 above)
Comment (laymonage) [reply client/src/controllers/SwapController.ts:278]: TypeScript's definition is `method?: string`, and `getAttribute()` returns `null` if the attribute doesn't exist. It will complain about being unable to assign `null` to `string | undefined`, and there are a few other ways around this:

- cast the value as `RequestInit['method']` anyway, suppressing TypeScript's complain but actually passing `method: null` in the options for `fetch`
- use `|| 'get'` instead, which is the default but we'll explicitly pass it instead of leaving it undefined
- use `...(formMethod ? { method: formMethod } : {})` or extract the RequestInit into a variable and do `if (formMethod) { init.method = formMethod; }`

Pick your poison? 😂

File: wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html (hunk 11 above)
Comment (laymonage) [reply wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html]: I've added a commit that uses `toggle_describedby` instead.

File: wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html (hunk 8 above)
Comment (laymonage) [reply wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html]: Yeah, I decided to not add capfirst at all, removed them in a new commit

File: client/src/controllers/SwapController.ts (hunk 15 above)
Comment (laymonage) [reply client/src/controllers/SwapController.ts]: I've added a commit that handles the case when there are tooltips shown for an element in the target container

File: wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html (hunk 10 above)
Comment (thibaudcolas) [reply wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html:119]: Thank you, I had missed that

File: client/src/controllers/SwapController.ts (hunk 14 above)
Comment (thibaudcolas) [reply client/src/controllers/SwapController.ts:282]: 🤡

File: client/src/controllers/UnsavedController.ts (hunk 4 above)
Comment (laymonage) [reply client/src/controllers/UnsavedController.ts:188]: I've opened a new issue for this: #12132

File: client/scss/components/_editing-sessions.scss [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+  @apply w-p-0 w-bg-transparent;
+
+  [aria-expanded='true']
+    .w-editing-sessions__decorated-avatar
+    :where(.icon, .w-editing-sessions__decorator-filler) {
+    @apply w-hidden;
+  }
+}
+
+.w-editing-sessions__session--latest {
```
[the remark this answers is not in the document]
Comment (laymonage) [reply client/scss/components/_editing-sessions.scss:48]: It means the session of a user who has saved the latest revision, so I think `session--outdated` is incorrect.

Maybe `session--saved`?

File: client/src/controllers/ActionController.test.js [hunk 19]
```diff
@@ -86,6 +134,113 @@ describe('ActionController', () => {
...
+      <button
+        id="button"
+        data-controller="w-action"
+        data-action="click->w-action#forceReload"
+      >
+        Force reload
+      </button>`);
+    });
+
+    it('should reload the page without preventing the beforeunload event', () => {
```
[the remark this answers is not in the document]
Comment (laymonage) [reply client/src/controllers/ActionController.test.js:218]: `force-value` comes from `UnsavedController` and you are half-correct: it's to force the controller to think that there are unsaved changes. In real life, this is used after submitting an invalid form, so you immediately have the "unsaved changes" message at the bottom and the browser confirmation will be shown even if you don't make further edits since the page load (as it assumes your invalid changes are not yet saved).

Even when `force-value` is `true`, there's a legit use case where you want to show the "unsaved changes" message at the footer but also bypass the browser confirmation: when you submit the form. This is done by setting `confirmationValue` to an empty string:

https://github.com/wagtail/wagtail/blob/2d568dd825e4b087d7cea1c4d1fa49d916b395c2/client/src/controllers/UnsavedController.ts#L234-L241

The actual trigger is here:
https://github.com/wagtail/wagtail/blob/2d568dd825e4b087d7cea1c4d1fa49d916b395c2/client/src/controllers/UnsavedController.ts#L174-L191

You can see that `forceValue` basically is an override for `hasCommentsValue` and `hasEditsValue`, but it doesn't have complete control over whether the browser confirmation dialog is shown or not (as the `confirmationValue` takes priority). By the way, the way we trigger the browser confirmation really should be reworked (#12132).

TLDR: I'm just using `force-value` to avoid having to set up the "has edits" detection. I tried setting `has-edits-value` to `true` directly on the element, but apparently it clears it on `connect()` so that didn't work.

https://github.com/wagtail/wagtail/blob/2d568dd825e4b087d7cea1c4d1fa49d916b395c2/client/src/controllers/UnsavedController.ts#L98-L99

https://github.com/wagtail/wagtail/blob/2d568dd825e4b087d7cea1c4d1fa49d916b395c2/client/src/controllers/UnsavedController.ts#L166-L172

I guess I can try setting it after `connect()`, by doing `await Promise.resolve()` first.

File: client/src/controllers/ActionController.ts [hunk 20]
```diff
@@ -97,10 +99,47 @@ export class ActionController extends Controller<
...
+    return formElement;
+  }
+
+  post(event: Event) {
+    event.preventDefault();
+    event.stopPropagation();
+    const formElement = this.createFormElement();
     document.body.appendChild(formElement);
     formElement.submit();
   }
```
[the remark this answers is not in the document]
Comment (laymonage) [reply client/src/controllers/ActionController.ts:111]: The element is only appended in `post()`, which we don't use. We only use `sendBeacon()`, and the `createFormElement()` method only creates the element in memory, which I think would be cleaned up when the method finishes as it's never appended to the DOM.

File: client/scss/components/_editing-sessions.scss [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+  .w-editing-sessions__avatar {
+    @apply w-border-critical-200 w-ping w-ping--critical;
```
[the remark this answers is not in the document]
Comment (laymonage) [reply client/scss/components/_editing-sessions.scss]: You're right, I recall Ben mentioning he used a lighter shade of red for the dark theme.

I've changed this to use the `text-error` color.

File: client/src/controllers/ActionController.test.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (laymonage) [reply client/src/controllers/ActionController.test.js]: OK I've changed the test to use has-edits instead, and added a code comment as well.

File: wagtail/admin/templates/wagtailadmin/shared/headers/slim_header.html [hunk 22]
```diff
@@ -114,7 +114,13 @@ <h1 class="w-sr-only">
                 {% endif %}
             {% endblock %}
 
-            <div class="w-w-full sm:w-w-min w-flex sm:w-flex-nowrap sm:w-flex-row w-items-center w-p-0 sm:w-py-0 sm:w-pr-4 sm:w-justify-end">
+            <div class="w-w-full w-overflow-scroll sm:w-w-min w-flex sm:w-flex-nowrap sm:w-flex-row w-items-center w-p-0 sm:w-py-0 sm:w-pr-4 sm:w-justify-end">
```
Comment (thibaudcolas) [inline wagtail/admin/templates/wagtailadmin/shared/headers/slim_header.html:117]: `overflow: scroll` means scrollbar tracks are visible no matter what, so this is probably not what we want.

File: wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+                        <span class="w-editing-sessions__decorator-filler"></span>
+                    </div>
+                {% endfragment %}
+
+                {% if current_session.user == session.user %}
+                    {% blocktranslate trimmed asvar saved_new_version_message %}
+                        You saved a new version in another session
+                    {% endblocktranslate %}
+                {% else %}
+                    {% blocktranslate trimmed with user_name=session.user|user_display_name|default:_("System") asvar saved_new_version_message %}
```
Comment (thibaudcolas) [inline wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html:24]: In existing code we seem to be using `user_display_name` with no fallback. I don’t know how to arrive at a situation where the fallback would be used. Deleting the user’s account perhaps?

If we need this kind of replacement consistently I wonder if we should create a new filter that comes with the fallback.

File: wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+        {% endblocktranslate %}
+
+        {% fragment as more_sessions_toggle %}
+            +{{ sessions|length|add:"-4" }}
+        {% endfragment %}
+
+        <li>
+            <span id="w-editing-sessions-more-summary" class="w-sr-only">{{ num_other_users_message }}</span>
+
+            {% dropdown theme="drilldown" classname="w-editing-sessions__session w-editing-sessions__session--more" toggle_classname="w-editing-sessions__avatar" toggle_label=more_sessions_toggle toggle_describedby="w-editing-sessions-more-summary" %}
```
Comment (thibaudcolas) [inline wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html:92]: Excellent!

File: wagtail/admin/templates/wagtailadmin/shared/editing_sessions/module.html [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+    </div>
+{% enddialog %}
+
+{% blocktranslate trimmed with user_name=_("System") model_name=content_type.name asvar someone_has_saved_message %}
+    {{ user_name }} has saved a newer version of this {{ model_name }}
+{% endblocktranslate %}
+
+{% blocktranslate trimmed with user_name=_("System") asvar overwrite_message %}
+    Proceeding will overwrite the changes made by {{ user_name }}. Refreshing the page will lose any of your unsaved changes.
+{% endblocktranslate %}
```
Comment (thibaudcolas) [inline wagtail/admin/templates/wagtailadmin/shared/editing_sessions/module.html:16]: I don’t understand in what scenario those messages would display. Are there any scenarios where they’re _not_ overwritten by the ones in `EditingSessionsList`?

**Review by thibaudcolas — CHANGES_REQUESTED** [review]
Spotted two UI issues:

- The ping box-shadow is cropped by its container on the left. I’d recommend disabling the ping for now if time is too tight.
- The `overflow: scroll` causes scrollbar tracks to always appear. This needs a fix I think.

File: wagtail/admin/templates/wagtailadmin/shared/editing_sessions/module.html (hunk 25 above)
Comment (laymonage) [reply wagtail/admin/templates/wagtailadmin/shared/editing_sessions/module.html:16]: No, there aren't any intended scenarios (unless the dialog is triggered by custom code). I need to pass non-empty strings to the `{% dialog %}` tag so it would render the appropriate elements. I could change this to something like "Placeholder" if that works better?

File: wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html (hunk 23 above)
Comment (laymonage) [reply wagtail/admin/templates/wagtailadmin/shared/editing_sessions/list.html:24]: In the editing sessions case, this can happen if the new revision is made programmatically without a `user` object attached, or if the user was deleted after the revision was created.

**Review by thibaudcolas — APPROVED** [review]
Thank you @laymonage!

Tested in:

- Safari 17.4 macOS 14.4
- Chrome 126 macOS 14.4
- Firefox 128 macOS 14.4
- Firefox 120 Windows 11
- Safari iOS 17.5 
- Chrome Android 13

**Comment by lb-** [comment]
Just posting a wow and massive congrats to getting this in, it's great to see this build on the Stimulus work done but also... This work is huge!

Super excited to see this go out to the community.
